lib/gen_debouncer.ex

```diff
@@ -0,0 +1,169 @@
+    end
+
+    @spec clean_work(t(), list) :: t()
+    def clean_work(state, keys) do
+      %__MODULE__{state | waiting_list: Map.drop(state.waiting_list, keys)}
+    end
+  end
+
+  defmodule Fragment do
+    @moduledoc false
+
+    @enforce_keys [:last_updated, :payloads]
+    defstruct @enforce_keys
+
+    @type t :: %__MODULE__{
+            last_updated: DateTime.t(),
+            payloads: MapSet.t()
+          }
+
+    # This is needed to initialize datetime properly.
+    # If used struct defaults it would start with
+    # the same datetime from compilation time.
+    @spec new :: t()
+    def new do
+      %__MODULE__{
+        last_updated: DateTime.utc_now(),
+        payloads: MapSet.new()
+      }
+    end
+
+    @spec add_payload(t(), any) :: t()
+    def add_payload(fragment, payload) do
+      fragment.payloads
+      |> MapSet.put(payload)
+      |> replace_payloads(fragment)
+    end
+
+    @spec add_payloads(t(), any) :: t()
+    def add_payloads(fragment, payloads) do
+      payloads = MapSet.new(payloads)
+
+      fragment.payloads
+      |> MapSet.union(payloads)
+      |> replace_payloads(fragment)
+    end
+
+    defp replace_payloads(payloads, fragment) do
+      %__MODULE__{
+        fragment
+        | payloads: payloads
+      }
+    end
+
+    @spec stale?(t(), non_neg_integer()) :: boolean()
+    def stale?(fragment, interval) do
+      now = DateTime.utc_now()
+
+      DateTime.diff(now, fragment.last_updated) |> abs() > interval and
+        not Enum.empty?(fragment.payloads)
+    end
+  end
+
+  @callback work(key :: any(), payload :: [any()]) :: :ok
+
+  @spec start_link(module(), non_neg_integer(), Keyword.t()) ::
+          GenServer.on_start()
+  def start_link(worker, interval, opts \\ []) do
+    GenServer.start_link(__MODULE__, {worker, interval}, opts)
+  end
+
+  @spec schedule_work(key :: any(), key :: any(), payload :: any()) ::
+          :ok | {:error, term()}
+  def schedule_work(debouncer, key, payload) do
+    GenServer.cast(debouncer, {:schedule_work, payload, key})
+  end
+
+  @spec schedule_work_batch(key :: any(), key :: any(), payload :: [any()]) ::
+          :ok | {:error, term()}
+  def schedule_work_batch(debouncer, key, payloads) do
+    GenServer.cast(debouncer, {:schedule_work_batch, payloads, key})
+  end
+
+  # -----
+  # GenServer Api implementation
+  # -----
+
+  @impl GenServer
+  def init({worker, interval}) do
+    state = %State{
+      worker: worker,
+      interval: interval
+    }
+
+    schedule_tick(state)
+
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_cast({:schedule_work, payload, key}, state) do
+    {:noreply, State.add_work(state, key, payload)}
+  end
+
+  def handle_cast({:schedule_work_batch, payloads, key}, state) do
+    {:noreply, State.add_work_batch(state, key, payloads)}
+  end
+
+  @impl true
+  def handle_info(:tick, state) do
+    schedule_tick(state)
+    workload = State.work_to_schedule(state)
```

Comment:
What if instead of having a function that filters out work that is ready to be worked on and later you drop it from state, you could have a single function that pops the work immediately from state and returns the work ready to be worked and the updated state?

Just trying to save an iteration (or two) over the work items.

Reply:
You are technically correct (which is the best kind of correct). I didn't want complex return types, but might just do that.